### ISSUE TITLE
Resource that uses with_run_context does not trigger notifications

### ISSUE THREAD

**Issue by martinisoft** [chef/chef#6987]
## Description

If you write a custom resource that uses the accumulator pattern via `with_run_context` and `edit_resource` or `find_resource` the resulting resource does not trigger any notifications for the custom resource.

## Chef Version

Chef 12.16 & 13.8

## Platform Version
Tested on Ubuntu 14.04 and 16.04 with both of the above chef client versions.

## Replication Case

Create an accumulator, we'll call this one aliases:

```ruby
property :address, String
property :recipients, Array

action :add do
  with_run_context :root do
    edit_resource(:template, "/tmp/aliases") do |new_resource|
      source 'aliases.erb'
      variables[:aliases] ||= {}
      variables[:aliases][new_resource.address] ||= []
      variables[:aliases][new_resource.address] += new_resource.recipients
      action :nothing
      delayed_action :create
    end
  end
end
```

Then in a recipe, consume the resource and try to attach a service reload signal:

```ruby
service 'mailserver' do
  supports reload: true
  action :nothing
end

foo_aliases do
  address 'bar'
  recipients ['baz', 'quux']
  notifies :reload, 'service[mailserver]' 
end
```

After foo_aliases converges at the end of the chef run, the service will not be signaled at all.

## Workaround

A workaround does not exist (to my knowledge) as the sub-resources that are created via find/edit resource do not mark the returned resource as updated.

**Comment by coderanger** [comment]
I think this is the correct behavior. The resource you are editing it not running inside the action so no notifications should be generated. It will generate notifications in the context it actually runs in.

**Comment by thommay** [comment]
@coderanger it might be correct, but it's certainly surprising and also not what one would expect on seeing just the recipe code.

**Comment by jonlives** [comment]
Indeed, it feels like the resulting behaviour here is definitely surprising, even if it is as coded.

**Comment by coderanger** [comment]
I think we have all possibly just internalized the pattern a tad too much :) I think we should lean in the direction of seeing `edit_resource` should be seen as a "oh god what fresh hell is this"-level hack any time it is used (even if there isn't always a better option).

**Comment by thommay** [comment]
This is the way we're recommending people build accumulator-esque resources, so I don't think that's practical. Either: 
 * `edit_resource` is bad, and hence we should build an api to wrap up the accumulator dance as seen as bad - which requires planning, and we should really do for 14 and backport in to 13 ASAP - which can then bubble up notifications, or
 * `edit_resource` is here to stay with this pattern, in which case this is a bug that needs fixing.
Either way, there is a problem here that we need to deal with.

**Comment by coderanger** [comment]
I mean I don't think `edit_resource` is bad, but I think its use should indicate you are doing Advanced Things so that you are willing to take responsibility for these kinds of edge cases. I don't think the semantics here are simple enough that there is one answer. In this particular use case, the expected behavior is probably to signal `update?` on every accumulator resource if any is updated, but there are lots of use cases of `edit_resource` where that isn't the case. If we want to gold-plate the accumulator use case then yes, we should make an API for that in particular, but we've been trying to do that for years and the problem is still that everyone has a different idea of how accumulators should work.

**Comment by martinisoft** [comment]
After sleeping on it and thinking about it deeper, I realize the problem stems from how `edit_resource` handles managing the resource collection. The missing piece here to me seems to be something on the order of an interface. One that the resource can apply and attach notifications to _it's_ managed resources should you end up in this land of crazy that I ended up in. Something like a method that will interrogate the custom resource invocation for any attached notifications and allow you to re-apply those to a given sub-resource.

What I believe is happening here is that because edit_resource is manipulating a resource during execution, it's being shoved to the bottom of the stack and thus won't _actually_ execute and return `@updated` as true which would then trigger bubbling up of notifications.

**Comment by lamont-granquist** [comment]
i'm completely confused as to how the workaround actually works.  that `if t.updated` happens at compile time during the provider/action run so it doesn't know anything about the delayed action running or not.  that should always be false within the provider, if its ever true then something weird is going on and the accumulator resource is being fired eagerly within the provider and that means that you're not getting lazy accumulator behavior, which is probably buggy.

and this has nothing to do with the edit_resource / with_root_context API as far as i can tell.  what you want is some way of taking the notifications off of the custom resource and reattaching them to the edited resource.

the first thing i'd suggest is that since its an accumulator it makes zero sense at all to hang notifies off of a fraction of your state like that.  to be at all correct that means that every accumulator call will need to have the notification copypasta'd to it.  that suggests that the notification needs to be hung off of the edit_resource directly and just made 100% the responsibility of the accumulator code and not exposed to the consumer of that API.  if you really want that flexibility there's nothing stopping you from having a `property :notify_on_change, Hash` which takes `{ resource => action }` pairs and then have the edit_resource add the notifications to the resource.  i'd also suggest making sure that you had a hook to simply do injection of notifications without updating the aliases at all -- so either make sure the resource works in a "null" case without any aliases and just hanging notifications off of it, or else just make a separate 'accumulator' which took notification arguments and applied them to the accumulated resource.  really this is some kind of dependency injection API.

there's a possible feature request here for some way to pull the notifications off of the wrapping resource and apply them to the resource being built.  but doing that by default in any sort of general case would be insane magic since accumulators that themselves create multiple resources wouldn't want all the wrapping notifications applied to every accumulated sub resource.

generally, though, i think this is screaming out that its the wrong pattern and the user needs to not be responsible for that and this is going down to wrong path to tactically fix something that needs to be fixed with a different approach.  its possible that chef needs better DI support in core to better deal with this, but that would just render this a symptom of a larger problem that is in no way a bug with edit_resource and needs to be addressed at a much higher level ("poise in core or something else").

anyway TL;DR no bug in edit_resource here, really not the right way to do this, you can still do it with properties manually if you really really want to, and nothing we need to worry about with Chef-14.

**Comment by martinisoft** [comment]
I've removed my workaround as I realize it does not work. Doing a debug log of the resource objects upon inspection reveals that the returned resource from with_run_context and edit_resource don't change the updated flag which is what I am guessing is chef's trigger for notifications.

**Comment by lamont-granquist** [comment]
what we should probably do is allow custom resources to fire an exception and crash if any notifications are hung off of them, because hanging a notification off of an accumulator is semantically crazy to the end user:

```ruby
foo_aliases do
  address 'bar'
  recipients ['baz', 'quux']
  notifies :reload, 'service[bar]' 
end

foo_aliases do
  address 'baz'
  recipients ['baz', 'bob']
  notifies :reload, 'service[baz]' 
end
```

what should the end user expect that does, because it looks to me like if you update the bar address it only bounces bar, while if you update baz it bounces baz, and that is never, ever going to be supported by accumulators.  you'll get accumulated notifications and if any edit is done all the notifications will fire.

honestly think hanging notifications off accumulators is just "that's wrong don't do that".

**Comment by martinisoft** [comment]
> what should the end user expect that does, because it looks to me like if you update the bar address it only bounces bar, while if you update baz it bounces baz, and that is never, ever going to be supported by accumulators. you'll get accumulated notifications and if any edit is done all the notifications will fire.

Isn't the current behavior to only fire one notification even if multiple are queued?

**Comment by thommay** [comment]
> honestly think hanging notifications off accumulators is just "that's wrong don't do that".

I disagree - because you're still going to want to be able to hang a notification off the final state of the resource, for instance if you accumulate a bunch of config blocks in to a file you'll want to restart the service that reads that file at the end.

Of course there are some edge cases there (`:before` or `:immediate` should never be allowed, for starters), but users are going to expect this to work.

**Comment by lamont-granquist** [comment]
different notifications there though...   one is `service[bar]` one is `service[baz]`.

we will get bug reports that "i only hug the restart of service[bar] off of the resource that added bar, why did baz get restarted?"

**Comment by thommay** [comment]
> we will get bug reports that "i only hug the restart of service[bar] off of the resource that added bar, why did baz get restarted?"

and we will close them with "because you told us to do so"

**Comment by martinisoft** [comment]
> different notifications there though... one is service[bar] one is service[baz].

Ah I misread, but I still stand by my previous statement as signaling both doesn't seem like strange behavior to me. If I did this on a chef core resource it would still do the same thing.

**Comment by lamont-granquist** [comment]
if you look at iptables we just do that work for the user:

https://github.com/chef-cookbooks/iptables/blob/master/resources/rule.rb#L43

that's the common pattern to follow, where the accumulator is responsible for setting up its own notfiications.

similarly here if you update the aliases, you should always rebuild the aliases.db and i don't understand why that needs to be an interface.

if there's a better use to to highlight it, it might help a lot.

still this is not an edit_resource/with_root_context bug though.   at most it looks like an enhancement to `edit_resource(:template, "/tmp/aliases", apply_wrapping_notifications: true) do ...`

**Comment by martinisoft** [comment]
So should I write an RFC for this particular thing? 

In the mean time I'll offer up a doc patch to the official discussion about accumulators to note this particular quirk.

**Comment by aespinosa** [comment]
@martinisoft what is the criteria for the custom resource to say that it's updated? is it that it edited the resource it wraps? or the resource being accumulated was updated?

**Comment by martinisoft** [comment]
> @martinisoft what is the criteria for the custom resource to say that it's updated? is it that it edited the resource it wraps? or the resource being accumulated was updated?

The resource being accumulated was updated. Right now both methods in that block return the resource, but don't flip the updated bit and rightly so since it hasn't actually taken action yet. The resource is being fully assembled, _then_ it is executed. By this point there is no context for the accumulated resource to signal back to it's original caller that it updated or not. So I'd want a way for the accumulated resource to have a callback to signal to it's original parent that it changed and thus will queue any notifications originally attached.

**Comment by lamont-granquist** [comment]
editing the resource will happen on every run, it should happen every time the accumulator is called, so it doesn't seem like that would make much sense to me.  modifying the resource on the resource collection isn't modifying the system state.  that's why edit_resource isn't a resource itself, doesn't update the `updated_by_last_action` status and doesn't fire notifications on editing resources.  that's also consistent with the way that accumulators have always been written -- dating back to when you had to use definitions to implement them (although you didn't really, but there was a lot of ceremonial diving around in core chef you had to do).

**Comment by aespinosa** [comment]
in this case then wrapping the whole thing in `converge_by` should be enough since updating the resource being accumulated is always updated. There's no sense of idempotence in accumulation

```
converge_by  ... do
   with_run_context ...
...
...
  end
end
```

**Comment by coderanger** [comment]
@aespinosa Except there is, but you can't tell until all accumulated resources are checked against the system. Which means notifications in any individual one are at best complicated, which is why I don't think there is a single obvious semantic to use here.

**Comment by martinisoft** [comment]
> in this case then wrapping the whole thing in converge_by should be enough since updating the resource being accumulated is always updated. There's no sense of idempotence in accumulation

That works, but it'll signal every single time though regardless which is overkill. It's my understanding that converge_by will always trigger an update regardless of the block actually changing. While edit_resource does modify a resource the resulting resource may end up matching what is in the original attribute data and thus won't need an update.

**Comment by martinisoft** [comment]
This pattern:

```ruby
foo_aliases do
  address 'bar'
  recipients ['baz', 'quux']
  notifies :reload, 'service[bar]' 
end

foo_aliases do
  address 'bar'
  recipients ['quux', 'corge']
  notifies :reload, 'service[baz]' 
end
```

Results in:

```ruby
foo_aliases do
  address 'bar'
  recipients ['baz', 'quux', 'quux', 'corge']
  notifies :reload, 'service[bar]'
  notifies :reload, 'service[baz]'
end
```

And if the above has executed in a previous run, it should return false for updated since it has previously changed the resource.

**Comment by aespinosa** [comment]
i guess it entire depends on the accumulation code

```
     variables[:aliases] ||= {}
      variables[:aliases][new_resource.address] ||= []
```

Can probably be translated to some `if ...else converge_by` incantation.  `.     `variables[:aliases][new_resource.address] += new_resource.recipients`  is not idempotent at all and should always trigger an update then. some more idempotence check should probably be made there

**Comment by martinisoft** [comment]
> Notifications are already a callback queue though, with de-duplication. If you start calling back to notifications in prior resources then it feels like you're building meta-notification and will need to meta-de-dup to avoid bouncing services multiple times. You probably want to take the notifications on the accumulator call and apply them to the resource being built and let the notifications just deal with de-up and other features normally.

That is what I'm asking for. I want a way for the accumulated resource to eventually return a correct state of `updated_by_last_action` so the normal signaling applies from the parent custom resource. Otherwise having a helper method in the parent custom resource that can re-apply it's given signals down to the accumulated children then this will happen magically. We'd have to note that immediate or before notifications can't fire this way since the accumulated resource seems to get pushed to the bottom of the stack.

**Comment by martinisoft** [comment]
Something like `new_resource.notifications` which would go into the edit_resource block could work.

**Comment by lamont-granquist** [comment]
I'm certain you can hack it up with raw ruby to work right now...

**Comment by lamont-granquist** [comment]
more or less this approach, if you fix all the bugs i've certainly written:

```ruby
r = with_run_context { edit_resource { .... } } # this is the resource we're building
notifications = run_context.parent.delayed_notification_collection[new_resource.declared_key]  # get all the notifications off the new_resource
notifications = notifications.map { |n| Notification.new(n.resource, n.action, r) }  # convert them to notifications from the resource we're building (being careful about not mutating the existing notifications)
notifications.each do { |n| run_context.root.notifies_delayed(n) }  # apply to the root run context where we're building the accumulated resource
```

**Comment by martinisoft** [comment]
It just dawned on me that the probably cause of this insanity is `delayed_action` which is effectively making notifications async from the original context so it won't return until the very end of the run, but by then it's too late for the wrapping resource.

**Comment by lamont-granquist** [comment]
if you want your accumulator to be able to be called from other custom resources, though, then you need to delay the action until the end of the run, so that's the correct way to write a lazy/delayed accumulator.

there's another way of writing an accumulator which is an eager-compile-time approach that i used for `multipackage` but that suffers from the problem that it only accumulates in each parent context's compile mode.

so if you call `multipackage` in recipe code those all get merged into one.  if you have a custom resource that calls `multipackage` that is entirely separate and all the multipackage calls in that custom resource will get accumulated and will run once.  there's no merging across run_contexts or up the stack.  and in that case you use `with_run_context :parent` and you just accept that it can't be perfect.  the reason to do it that way is that most of the time what you want with packages is to coalesce them across recipe code and do one thunk very early in the run in order to install everything, and delayed installation of packages to the end of the run is going to be what nobody wants.  so it has to be eager in order to be at all useful, and the weird breakage from calling inside of child run_context's just has to be accepted.

for most accumulators they need to be delayed so that they can be called from custom resources in addition to being called from recipe code.

there's a crazy edge case of delayed notifications in the outer run context which call this kind of lazy accumulator which would construct a kind of arms race to lazy things to the end of the run, which thankfully i've never seen come up in practice (i.e. if you sent a delayed notification to a foo_aliases call from the outer run context, that would be a bit crazy...)

**Comment by lamont-granquist** [comment]
And a sort of pseudo-diagram:

```
*  outer run context's compile mode runs
*  outer run context's converge mode starts
    * some internal chef resources might run
    * the first custom resources compile mode runs
    * the first custom resources converge mode runs
    * [....]
    * the last custom resources compile mode runs
    * the last custom resources converge mode runs
    * maybe a couple more internal chef resources run
* outer run_context's delayed notifications fire
```

if that last custom_resource's provider could have a `foo_aliases` call in it, then you need to delay the editing of the template to the delayed notifications phase, or else at some point in the run you'll write out a partial template, then you'll call that last custom_resource with the `foo_aliases` call in it, and even if you `:create` the template again you'll thrash the system state.   the only place that you can be ensured that all the provider code in all the custom resources in the run have complete, so all the accumulated state is accumulated, is during delayed notifications.

**Comment by lamont-granquist** [comment]
Closing this as it is way too much work for way too much of an edge case and my recommendation is to never do this.  I haven't seen a compelling argument that this is the only way to get the work done.
